### ISSUE TITLE
Memory leak in particles.v example

### ISSUE THREAD

**Issue by ArcOfDream** [vlang/v#11460]
**V version:** V 0.2.4 be0c54c
**OS:** Arch Linux

**What did you do?**
Run the particles.v example from the main repository. Tried with `-autofree` and `-gc boehm`, too.

**What did you expect to see?**
Particles!
 
**What did you see instead?**
Particles, but after a short while of moving a mouse around the whole system slowed to a crawl.

**Comment by spytheman** [comment]
For me, it leaks currently very fast with `-autofree` and without any options (2-3GB per second) for some reason.
With `-gc boehm` it stabilized at 204MB RSS, and does not leak further:
![image](https://user-images.githubusercontent.com/26967/132901602-c79d72e3-4b48-4d7d-8fa8-ac4266f54462.png)

I think this may be due to a recent change, since I remember running it with -autofree a 2-3 months ago, without leaks 🤔 .

**Comment by GZGavinZhao** [comment]
@spytheman  Through bisect, commit 4198205 is the culprit:
```
3f1482a31e69ad759a0be1ac5bd5939676fa452d is the first bad commit

commit 3f1482a31e69ad759a0be1ac5bd5939676fa452d
Author: vlang-bot <vlang-bot@users.noreply.github.com>
Date:   Wed Jul 21 19:01:34 2021 +0000

    [v:master] 4198205 - builtin: make `array.delete()` reallocate, provide `.delete_many()` (#10889)

 v.c     | 28 ++++++++++++++++++++--------
 v_win.c | 28 ++++++++++++++++++++--------
 2 files changed, 40 insertions(+), 16 deletions(-)
```

**Comment by GZGavinZhao** [comment]
Here's the problem, the `old_data` that is used to temporarily store the copy of the array is not being freed.
```v
// delete_many deletes `size` elements beginning with index `i`
pub fn (mut a array) delete_many(i int, size int) {
	$if !no_bounds_checking ? {
		if i < 0 || i + size > a.len {
			endidx := if size > 1 { '..${i + size}' } else { '' }
			panic('array.delete: index out of range (i == $i$endidx, a.len == $a.len)')
		}
	}
	// NB: if a is [12,34], a.len = 2, a.delete(0)
	// should move (2-0-1) elements = 1 element (the 34) forward
	old_data := a.data
	new_size := a.len - size
	new_cap := if new_size == 0 { 1 } else { new_size }
	a.data = vcalloc(new_cap * a.element_size)
	unsafe { vmemcpy(a.data, old_data, i * a.element_size) }
	unsafe {
		vmemcpy(&byte(a.data) + i * a.element_size, &byte(old_data) + (i + size) * a.element_size,
			(a.len - i - size) * a.element_size)
	}
	a.len = new_size
	a.cap = new_cap

        unsafe { free(old_data) } // FREE THIS!
}
```
@spytheman Is this enough for a pull request? I'm not sure if this is a problem with the function or if it is for `-autofree`, because this fix seems, perhaps, too "trivial"?

**Comment by spytheman** [comment]
> because this fix seems, perhaps, too "trivial"?

Unfortunately, it is. The problem is that slices share their memory, until it is modified, and so freeing the old data, will lead to invalidating data, that is potentially used by another slice. It can be seen by running vlib/builtin/array_test.v .

**Comment by ArcOfDream** [comment]
It seems that as of V 0.2.4 03269f9, the leak is gone with the `-gc boehm -autofree` flag. With just `-autofree` there's still a leak, and with just `-gc boehm` it just gives the following warning repeated some times
```
GC Warning: Repeated allocation of very large block (appr. size 319488):
	May lead to memory leak and poor performance```
But with both flags everything's OK. :+1:

**Comment by Congyuwang** [comment]
Just a question. Is V memory management like: 'finding a memory bug, fix it', instead of having a logical guarantee that the memory model works fine?

**Comment by GZGavinZhao** [comment]
As of 03/27/2022, seems to be fixed on the latest version c7a9242. I had to use `-cc clang` to compile because I'm getting the following building error when using `-cc tcc` and `-cc gcc`:
```
builder error: In file included from /<directories>/v/thirdparty/stb_image/stbi.c:3:
In file included from /<directories>/v/thirdparty/stb_image/stb_image.h:344:
In file included from /usr/include/stdio.h:33:
/usr/include/stdio.h:33: error: include file 'stddef.h' not found
```

Nevertheless, now even without using `-autofree`, there's no memory leak and the usage stays at around 22 MB. Someone needs to test this with `-cc tcc`.

**Comment by JalonSolov** [comment]
I didn't have any trouble compiling with tcc or gcc, but it does show a identical small leak (8,288 bytes) with both:
```
$ v examples/sokol/particles/particles.v
$ valgrind examples/sokol/particles/particles
==967508== Memcheck, a memory error detector
==967508== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==967508== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==967508== Command: examples/sokol/particles/particles
==967508== 
==967508== 
==967508== HEAP SUMMARY:
==967508==     in use at exit: 95,057,590 bytes in 1,375 blocks
==967508==   total heap usage: 32,791 allocs, 31,416 frees, 607,351,393 bytes allocated
==967508== 
==967508== LEAK SUMMARY:
==967508==    definitely lost: 8,288 bytes in 27 blocks
==967508==    indirectly lost: 160,115 bytes in 886 blocks
==967508==      possibly lost: 0 bytes in 0 blocks
==967508==    still reachable: 94,889,187 bytes in 462 blocks
==967508==         suppressed: 0 bytes in 0 blocks
==967508== Rerun with --leak-check=full to see details of leaked memory
==967508== 
==967508== For lists of detected and suppressed errors, rerun with: -s
==967508== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
$ v -cc gcc examples/sokol/particles/particles.v
$ valgrind examples/sokol/particles/particles
==967561== Memcheck, a memory error detector
==967561== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==967561== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==967561== Command: examples/sokol/particles/particles
==967561== 
==967561== 
==967561== HEAP SUMMARY:
==967561==     in use at exit: 95,057,590 bytes in 1,376 blocks
==967561==   total heap usage: 34,421 allocs, 33,045 frees, 551,246,628 bytes allocated
==967561== 
==967561== LEAK SUMMARY:
==967561==    definitely lost: 8,288 bytes in 28 blocks
==967561==    indirectly lost: 160,115 bytes in 886 blocks
==967561==      possibly lost: 0 bytes in 0 blocks
==967561==    still reachable: 94,889,187 bytes in 462 blocks
==967561==         suppressed: 0 bytes in 0 blocks
==967561== Rerun with --leak-check=full to see details of leaked memory
==967561== 
==967561== For lists of detected and suppressed errors, rerun with: -s
==967561== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
$
```

**Comment by GZGavinZhao** [comment]
Oh I wasn't using Valgrind because if I use that, the application gets too laggy. It's very possible that this leak also happens on Clang, but I'm unable to test it.

**Comment by medvednikov** [comment]
Fixed in 0.3.
